drivers/watchdog/wdt_cc13xx_cc26xx.c

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (c) 2022 Florin Stancu <niflostancu@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_cc13xx_cc26xx_watchdog
+
+#include <zephyr/drivers/watchdog.h>
+#include <soc.h>
+#include <errno.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wdt_cc13xx_cc26xx);
+
+/* Driverlib includes */
+#include <inc/hw_types.h>
+#include <inc/hw_wdt.h>
+#include <driverlib/rom.h>
+#include <driverlib/prcm.h>
+#include <driverlib/watchdog.h>
+
+
+/*
+ * TI CC13xx/CC26xx watchdog is a 32-bit timer that runs on the MCU clock
+ * with a fixed 32 divider.
+ *
+ * For the default MCU frequency of 48MHz:
+ * 1ms = (48e6 / 32 / 1000) = 1500 ticks
+ * Max. value = 2^32 / 1500 ~= 2863311 ms
+ *
+ * The watchdog will issue reset only on second in turn time-out (if the timer
+ * or the interrupt aren't reset after the first time-out). By default, regular
+ * interrupt is generated but platform supports also NMI (can be enabled by
+ * setting the `interrupt-nmi` boolean DT property).
+ */
+
+#define CPU_FREQ DT_PROP(DT_PATH(cpus, cpu_0), clock_frequency)
+#define WATCHDOG_DIV_RATIO	 32
+#define WATCHDOG_MS_RATIO	 (CPU_FREQ / WATCHDOG_DIV_RATIO / 1000)
+#define WATCHDOG_MAX_RELOAD_MS	 (0xFFFFFFFFu / WATCHDOG_MS_RATIO)
+#define WATCHDOG_MS_TO_TICKS(_ms) ((_ms) * WATCHDOG_MS_RATIO)
+
+struct wdt_cc13xx_cc26xx_data {
+	uint32_t reload;
+	wdt_callback_t cb;
+	uint8_t flags;
+};
+
+struct wdt_cc13xx_cc26xx_cfg {
+	uint32_t reg;
+	uint8_t irq_nmi;
+	void (*irq_cfg_func)(void);
+};
+
+static int wdt_cc13xx_cc26xx_install_timeout(const struct device *dev,
+				      const struct wdt_timeout_cfg *cfg)
+{
+	struct wdt_cc13xx_cc26xx_data *data = dev->data;
+
+	/* window watchdog not supported */
+	if (cfg->window.min != 0U || cfg->window.max == 0U) {
+		return -EINVAL;
+	}
+	/*
+	 * Note: since this SoC doesn't define CONFIG_WDT_MULTISTAGE, we don't need to
+	 * specifically check for it and return ENOTSUP
+	 */
+
+	if (cfg->window.max > WATCHDOG_MAX_RELOAD_MS) {
+		return -EINVAL;
+	}
+	data->reload = WATCHDOG_MS_TO_TICKS(cfg->window.max);
+	data->cb = cfg->callback;
+	data->flags = cfg->flags;
+	LOG_DBG("raw reload value: %d", data->reload);
+	return 0;
+}
+
+static int wdt_cc13xx_cc26xx_setup(const struct device *dev, uint8_t options)
+{
+	const struct wdt_cc13xx_cc26xx_cfg *config = dev->config;
+	struct wdt_cc13xx_cc26xx_data *data = dev->data;
+
+	/*
+	 * Note: don't check if watchdog is already enabled, an application might
```

Comment:
I have some doubts here, see: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/watchdog.h#L161

We allow to `disable` this watchdog (but only by disabling generating the reset) so the `setup` can be called again. But is this a correct approach? In fact, this type of watchdog cannot be stopped.

Reply:
Yeah, the primary use case for this was to be able change the timeout dynamically.
A Zephyr Watchdog API behavior change like being able to call `install_timeout` after setup would be a better way, though...

Reply:
Or, we could further _emulate_ the watchdog being stopped by defining a boolean flag that gets set when `wdt_disable` is called then preventing the execution of the user callback (note: the interrupt handler will still fire).